### PR TITLE
Replace non-generic WeakReference usage in several libraries

### PR DESCRIPTION
**Pull request by stephentoub**
The System.Private.Runtime.InteropServices.JavaScript changes are sufficient to let WeakReference be trimmed out of a default Blazor wasm app.

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlElementList.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ internal class XmlElementList : XmlNodeList
         private bool _atomized;     //whether the localname and namespaceuri are atomized
         private int _matchCount;   // cached list count. -1 means it needs reconstruction
 
-        private WeakReference? _listener;   // XmlElementListListener
+        private WeakReference<XmlElementListListener>? _listener;   // XmlElementListListener
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlElementList.cs:25]: (nit) is the comment necessary anymore?

**Review by eerhardt — APPROVED** [review]
LGTM - just one silly nit.

**Comment by stephentoub** [comment]
@kjpou1 or @steveisok, would you mind reviewing the JavaScript tweaks? (Or if there's someone better, loop them in.) Thanks!

**Comment by danmoseley** [comment]
Should we have an analyzer to discourage the non generic one?

**Comment by steveisok** [comment]
> @kjpou1 or @steveisok, would you mind reviewing the JavaScript tweaks? (Or if there's someone better, loop them in.) Thanks!

/cc @lewing

**Comment by stephentoub** [comment]
> Should we have an analyzer to discourage the non generic one?

We... could.  It'd kind of be like having an analyzer to discourage use of `ArrayList` and point folks to `List<T>`.  If we want an analyzer that suggests folks move from non-generics to their generic counterparts, we could, though it's likely to be quite noisy.

**Comment by danmoseley** [comment]
I guess I was thinking of our own codebase, but code reviews are enough for that.

**Review by lewing — APPROVED** [review]

**Pull request merged**